src/rich-text/commands/index.ts

```diff
@@ -141,6 +145,55 @@ function getHeadingLevel(state: EditorState): number {
     return level;
 }
 
+/**
+ * Creates a command that toggles tagLink formatting for a node
+ * @param allowNonAscii Whether to allow non-ascii characters in the tag name
+ */
+export function toggleTagLinkCommand(allowNonAscii: boolean) {
+    return (state: EditorState, dispatch?: (tr: Transaction) => void) => {
+        if (state.selection.empty) {
+            return false;
+        }
+
+        if (
+            !validateTagName(
+                state.selection
+                    .content()
+                    .content.firstChild?.textContent.trim(),
+                allowNonAscii
+            )
+        ) {
+            return false;
+        }
+
+        if (!dispatch) {
+            return true;
+        }
+
+        let tr = state.tr;
+        const nodeCheck = nodeTypeActive(state.schema.nodes.tagLink);
+        if (nodeCheck(state)) {
+            const selectedText = state.selection.content().content.firstChild
```

Comment:
This fails when the tag is not the first child in the slice, most notably, when a tag falls in the center of other text:

https://user-images.githubusercontent.com/14169651/178818641-1c43eda6-b35f-48c0-8ca3-a5aef92542a7.mp4

Reply:
The issue here was trying to validate the selected text as a tagName every time we were toggling, when we should only do so when we're about to create a tagLink. Moved the validation logic down a level and it works properly now.